### PR TITLE
Bootloop detection platform factoring

### PR DESCRIPTION
**Pull request by willmmiles**
- Change ESP8266 implementation to use direct RTC memory references, like the ESP32 implementation
- Move ESP8266 RTC variables out of region used by OTA Updater class
- Factor out the platform-specific code so the common logic is platform agnostic
- Reframe if-else chains with switch statements for clarity and reducing duplication

### CODE REVIEW

**Comment by DedeHai** [comment]
Cant review in detail or test rn, but looks like solid changes

File: wled00/util.cpp [hunk 1]
```diff
@@ -720,125 +720,131 @@ void *realloc_malloc(void *ptr, size_t size) {
...
+      //restoreConfig(); // TODO: blindly restoring config if brownout detected is a bad idea, need a better way (if at all)
+      break;
   }
-#endif
-  return false; // no bootloop detected
+  
+  bl_last_boottime = rtctime; // store current runtime for next reboot
+
+  return result;
 }
```
[the remark this answers is not in the document]
Comment (DedeHai) [reply wled00/util.cpp]: @willmmiles this is a fair point by the rabbit. if a system is somewhat unstable or there is a strange bug that would cause say two consecutive crashes every 24 hours, it will still detect a bootloop. In such a hypothetical scenario we would not want to reset the config as the user has enough time to take action. Adding an `else if` to reset the counter if the system has run for some time, say 5 minutes (?), would prevent seemingly random config resets in such edge cases.

**Comment by DedeHai** [comment]
code looks good, do you want me to run any tests?

File: wled00/util.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (willmmiles) [reply wled00/util.cpp]: No objections from me; it just wasn't part of your original logic.

Actually with that in mind - that a "boot loop" means that "I can't log in and change the config" -- I might also suggest increasing `BOOTLOOP_INTERVAL_MILLIS` to a much bigger number (2 minutes maybe?).  Wifi registration is sometimes pretty slow, and it can be tough to know when it's ready to accept requests.   Sometimes my test boards seem to fail to connect to the wifi the first time, have to wait for something to time out, then it works the second try.

**Comment by willmmiles** [comment]
> code looks good, do you want me to run any tests?

Any testing is always welcome!   I did validate that the variables worked properly with my ESP8266 test board last week.  Unfortunately I blew up its voltage regulator last night - now it's somewhat frankensteined together with a different unit where the CPU is blown, so the soft reset doesn't work and it can only OTA until I swap the chips, so I don't want to boot loop it for the moment.  New test board should arrive today.  (So nice they're so inexpensive!)

File: wled00/util.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (DedeHai) [reply wled00/util.cpp]: Yea, I did not think of that scenario when I wrote the code, the rabbit is kind of smart ;)
I am not sure there are no drawbacks by raising it to 2min. I mean, if something is wrong with the system or config to make it crash, I expect it to crash within a short amount of time after boot. But you are right, if the crash happens only after a wifi connection is established, raising the interval makes sense. I can't come up with any reasoning not to do that.

**Comment by willmmiles** [comment]
> `746-756`: **Defensive null-check for system_get_rst_info()**
> 
> Add a null-check to avoid a rare deref crash in early-boot edge cases.

This one's for you, @coderabbitai: `system_get_rst_info()` can never return null - it returns a pointer to static memory.  This is a fairly common technique in C APIs to provide an interface that hides the internal details of the placement of the structure, which might be part of some larger struct that is not meant to be exposed.

File: wled00/util.cpp [hunk 2]
```diff
@@ -720,125 +720,139 @@ void *realloc_malloc(void *ptr, size_t size) {
...
+
+#if ESP_IDF_VERSION >= ESP_IDF_VERSION_VAL(4, 4, 0)
+static inline uint32_t getRtcMillis() { return esp_rtc_get_time_us() / 1000;  }
+#elif ESP_IDF_VERSION >= ESP_IDF_VERSION_VAL(3, 3, 0)
+static inline uint32_t getRtcMillis() { return rtc_time_slowclk_to_us(rtc_time_get(), rtc_clk_slow_freq_get_hz()) / 1000; }
+#endif
+
 void bootloopCheckOTA() { bl_actiontracker = BOOTLOOP_ACTION_OTA; } // swap boot image if bootloop is detected instead of restoring config
+
 #endif
```
[the remark this answers is not in the document]
Comment (willmmiles) [reply wled00/util.cpp:781]: I'd considered it; honestly doing it that way would generally be my preferred style.  Unfortunately, given that the body and caller are in separate translation units, the empty call cannot be inlined away, so the result would be a small waste of limited code memory bytes on ESP8266 systems for nothing but code aesthetic reasons.  Given that the extra #ifdefs are already in the code, the best value is to leave them alone.

**Review by DedeHai — APPROVED** [review]

**Comment by DedeHai** [comment]
Tested on ESP8266 and C3, works as intended. I took the liberty to remove whitespaces and add a debug output if a crash is detected.

**Comment by DedeHai** [comment]
damn, forgot to push the commit beofre merge :) ok, will add it to main directly
